### DISCUSSION TITLE
List of Keras (images) Preprocessing Layer (KPL)

### DISCUSSION THREAD

**Discussion by innat** [keras-team/keras-cv#93] [General]
## Keras Preprocessing Layer (KPL)

A list to track down the image preprocessing layers, filter out from [issues](https://github.com/keras-team/keras-cv/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc). Please, feel free to share the relevant content for KPL. 

- [x] CutMix, MixUp augmentation
     - [x] [Classification](https://github.com/keras-team/keras-cv/issues/19) : [PR](https://github.com/keras-team/keras-cv/pull/24)
     - [ ] [Detection](https://github.com/keras-team/keras-cv/issues/35) : [PR](https://github.com/keras-team/keras-cv/pull/282) (WIP)
- [ ] Mosaic Augmentation - Object Detection
     - [ ] [Classification](https://github.com/keras-team/keras-cv/issues/250)
     - [ ] [Detection](https://github.com/keras-team/keras-cv/issues/21)
- [ ] RandomFlip
     - [x] [Classification](https://keras.io/api/layers/preprocessing_layers/image_augmentation/random_flip/)
     - [ ] Detection 
- [x] [CutOut](https://github.com/keras-team/keras-cv/issues/39) and [RandomErasing](https://github.com/keras-team/keras-cv/issues/27)
- [x] [Equalize](https://github.com/keras-team/keras-cv/issues/82)
- [x] [AutoContrast](https://github.com/keras-team/keras-cv/issues/81)
- [x] [GridMask](https://github.com/keras-team/keras-cv/issues/31)
- [ ] [SLIC Layer: Superpixel](https://github.com/keras-team/keras-cv/issues/43)
- [x] [RandAug](https://github.com/keras-team/keras-cv/issues/29) : [PR](https://github.com/keras-team/keras-cv/pull/281)
- [ ] [AutoAugment](https://github.com/keras-team/keras-cv/issues/41)
- [x] [AugMix](https://github.com/keras-team/keras-cv/issues/40)
- [x] [Solarize](https://github.com/keras-team/keras-cv/issues/90) : [PR](https://github.com/keras-team/keras-cv/pull/106)
- [x] [Shear](https://github.com/keras-team/keras-cv/issues/89) : [PR](https://github.com/keras-team/keras-cv/pull/180)
- [x] [Sharpen](https://github.com/keras-team/keras-cv/issues/88)
- [x] [RandomBrightness](https://github.com/keras-team/keras-cv/issues/87)
- [x] [Color degeneration](https://github.com/keras-team/keras-cv/issues/86) : [PR](https://github.com/keras-team/keras-cv/pull/205)
- [x] [Posterize](https://github.com/keras-team/keras-cv/issues/84)
- [x] [Invert](https://github.com/keras-team/keras-cv/issues/83)
- [ ] [Simple Copy-Paste - Instance Segmentation](https://github.com/keras-team/keras-cv/issues/28)
- [x] [Color Jitter](https://pytorch.org/vision/stable/auto_examples/plot_transforms.html#colorjitter) : [PR](https://github.com/keras-team/keras-cv/pull/158)
- [x] [Grayscale](https://pytorch.org/vision/stable/auto_examples/plot_transforms.html#grayscale) : [PR](https://github.com/keras-team/keras-cv/pull/132)
- [x] [RGBShift](https://albumentations.ai/docs/api_reference/augmentations/transforms/#albumentations.augmentations.transforms.RGBShift) : [PR](https://github.com/keras-team/keras-cv/pull/128)
 - [x] [ChannelShuffle](https://albumentations.ai/docs/api_reference/augmentations/transforms/#albumentations.augmentations.transforms.ChannelShuffle) : [PR](https://github.com/keras-team/keras-cv/pull/124)
- [x] [Hue Value](https://github.com/keras-team/keras-cv/pull/231)
- [x] [Saturation Value](https://github.com/keras-team/keras-cv/pull/221)

**Comment by innat** [comment]
From [keras.io](https://keras.io/)
- [Image preprocessing layers](https://keras.io/api/layers/preprocessing_layers/image_preprocessing/)
- [Image augmentation layers](https://keras.io/api/layers/preprocessing_layers/image_augmentation/)

**Comment by bhack** [comment]
One of these days we will probably need to care about how to handle classes in our "canned policies":

https://arxiv.org/abs/2204.03632.

**Comment by innat** [comment]
`KerasCV` :face_with_spiral_eyes:

![image](https://user-images.githubusercontent.com/17668390/169665933-a2f3f6b4-5d61-447c-8b4a-c72652a08ecc.gif)

**Reply by LukeWood** [reply]
What an awesome Gif!!!!!!  Would you share the script you used to create it?

**Reply by LukeWood** [reply]
> `KerasCV` 😵‍💫
> 
> ![image](https://user-images.githubusercontent.com/17668390/169665933-a2f3f6b4-5d61-447c-8b4a-c72652a08ecc.gif) [ ](https://user-images.githubusercontent.com/17668390/169665933-a2f3f6b4-5d61-447c-8b4a-c72652a08ecc.gif) ![image](https://user-images.githubusercontent.com/17668390/169665933-a2f3f6b4-5d61-447c-8b4a-c72652a08ecc.gif)

I think your RandAugment is using an old bugged version.  You may want to upgrade KerasCV again!

**Reply by innat** [reply]
> What an awesome Gif!!!!!! Would you share the script you used to create it?

haha, thanks. I made it on last week for fun. 😅
Though I didn't save the script I will share a new one.

**Reply by LukeWood** [reply]
> > What an awesome Gif!!!!!! Would you share the script you used to create it?
> 
> haha, thanks. I made it on last week for fun. 😅 Though I didn't save the script I will share a new one.

Dang!  It’s a really cool visual.  Perhaps a good addition to our marketing materials if CutMix + MixUp + fixed RandAugment are included 👀

**Reply by bhack** [reply]
Nice, but I suggest to change the image in marketing materials:

https://github.com/scikit-image/scikit-image/issues/1855

**Reply by innat** [reply]
@LukeWood 
Here are the scripts. [COLAB](https://colab.research.google.com/drive/1cKAM7R7I2GO6OzfkxwLH6TTpXL5-_q7u?usp=sharing).

@bhack 
Yes, and `data.astronaut()` would be a nice choice. The official pytorch tutorials also use it.

**Reply by LukeWood** [reply]
Thank you @innat!  I might produce one that uses CutMix and MixUp too and add it to our marketing :)

I'll be sure to reference you in the final version!  Such a cool idea.

**Comment by bhack** [comment]
[RegMixup: Mixup as a Regularizer Can Surprisingly Improve Accuracy and Out Distribution Robustness](https://arxiv.org/abs/2206.14502)
